File: apps/gui-flutter/lib/mixer/pad_modes.dart
import 'package:flutter/services.dart';
import 'package:gui_flutter/src/rust/api/engine.dart' show PadMode;

export 'package:gui_flutter/src/rust/api/engine.dart' show PadMode;

/// Pad mode helpers + Tauri-matching beat tables / labels.

const kPadModes = <PadMode>[
  PadMode.hotCue,
  PadMode.loopRoll,
  PadMode.beatJump,
  PadMode.sampler,
];

const kLoopRollBeats = <num>[1 / 32, 1 / 16, 1 / 8, 1 / 4, 1 / 2, 1, 2, 4];
const kBeatJumpForward = <num>[1, 2, 4, 8, 16, 32, 64, 128];
const kBeatJumpBack = <num>[-1, -2, -4, -8, -16, -32, -64, -128];

/// Sampler play-mode wire values (Tauri `SamplerPlayMode`).
const kSamplerPlayModeOneshot = 'oneshot';
const kSamplerPlayModeHold = 'hold';
const kSamplerPlayModeLoop = 'loop';

/// Default sampler play mode (Tauri `DEFAULT_SAMPLER_PLAY_MODE`).
const kDefaultSamplerPlayMode = kSamplerPlayModeOneshot;

/// Bank settings dialog options (`default` = inherit settings).
const kSamplerPlayModeOptions = <String>[
  'default',
  kSamplerPlayModeOneshot,
  kSamplerPlayModeHold,
  kSamplerPlayModeLoop,
];

String padModeShortLabel(PadMode mode) => switch (mode) {
  PadMode.hotCue => 'Cue',
  PadMode.loopRoll => 'Roll',
  PadMode.beatJump => 'Jump',
  PadMode.sampler => 'Sample',
};

PadMode cyclePadMode(PadMode mode, int direction) {
  final index = kPadModes.indexOf(mode);
  final current = index >= 0 ? index : 0;
  final len = kPadModes.length;
  final next = direction < 0 ? (current + len - 1) % len : (current + 1) % len;
  return kPadModes[next];
}

bool shiftKeyPressed() =>
    HardwareKeyboard.instance.isLogicalKeyPressed(
      LogicalKeyboardKey.shiftLeft,
    ) ||
    HardwareKeyboard.instance.isLogicalKeyPressed(
      LogicalKeyboardKey.shiftRight,
    );

File: apps/gui-app/src/lib/pad-modes.ts
import type { PadMode } from "@/types";

export const PAD_MODES: readonly PadMode[] = [
  "hot_cue",
  "loop_roll",
  "beat_jump",
  "sampler",
] as const;

export function cyclePadMode(mode: PadMode, direction: number): PadMode {
  const index = PAD_MODES.indexOf(mode);
  const current = index >= 0 ? index : 0;
  const len = PAD_MODES.length;
  const next = direction < 0 ? (current + len - 1) % len : (current + 1) % len;
  return PAD_MODES[next] ?? "hot_cue";
}

export const PAD_MODE_LABELS: Record<PadMode, string> = {
  hot_cue: "Hot Cue",
  loop_roll: "Loop Roll",
  beat_jump: "Beat Jump",
  sampler: "Sampler",
};

export const PAD_MODE_SHORT_LABELS: Record<PadMode, string> = {
  hot_cue: "Cue",
  loop_roll: "Roll",
  beat_jump: "Jump",
  sampler: "Sample",
};

export const LOOP_ROLL_BEATS = [1 / 32, 1 / 16, 1 / 8, 1 / 4, 1 / 2, 1, 2, 4] as const;
export const BEAT_JUMP_FORWARD = [1, 2, 4, 8, 16, 32, 64, 128] as const;
export const BEAT_JUMP_BACK = [-1, -2, -4, -8, -16, -32, -64, -128] as const;

export function formatBeatLength(beats: number): string {
  if (beats >= 1 || beats <= 0) {
    return String(beats);
  }
  return `1/${Math.round(1 / beats)}`;
}
